server/resp/src/main/java/org/infinispan/server/resp/operation/SetOperation.java

```diff
@@ -0,0 +1,212 @@
+         this.key = key;
+      }
+
+      public void withValue(byte[] value) {
+         this.value = value;
+      }
+
+      public void withReturnPrevious() {
+         this.setAndReturnPrevious = true;
+      }
+
+      public void withExpiration(long expirationMs) {
+         this.expirationMs = expirationMs;
+      }
+
+      public void withOperationType(SetOperationType operationType) {
+         this.operationType = operationType;
+      }
+
+      public void withKeepTtl() {
+         this.keepTtl = true;
+      }
+
+      public boolean isKeepingTtl() {
+         return this.keepTtl;
+      }
+
+      public boolean isReturningPrevious() {
+         return this.setAndReturnPrevious;
+      }
+
+      public SetOperationType typeOfOperation() {
+         return this.operationType;
+      }
+
+      public boolean isUsingExpiration() {
+         return expirationMs > 0 || isKeepingTtl();
+      }
+
+      private void parseAndLoadOptions() {
+         Iterator<byte[]> it = arguments.iterator();
+         while (it.hasNext()) {
+            if (key == null) {
+               withKey(it.next());
+               continue;
+            }
+
+            if (value == null) {
+               withValue(it.next());
+               continue;
+            }
+
+            // Bellow here we parse the optional arguments for the SET command:
+            //
+            // * `GET`: return the previous value with this key or nil;
+            // * `NX` or `XX`: putIfAbsent or putIfPresent. Returns nil if failed, if `GET` is present, it takes precedence.
+            //
+            // And expiration related parameters:
+            //
+            // `EX` seconds: TTL in seconds;
+            // `PX` milliseconds: TTL in ms;
+            // `EXAT` timestamp: Unix time for key expiration, seconds;
+            // `PXAT` timestamp: Unix time for key expiration, milliseconds;
+            // `KEEPTTL`: keep the key current TTL.
+            //
+            // Each of the time arguments are exclusive, only one is present at a time.
+            // All these arguments can be in any order. Expiration must be followed by the proper value.
+            String arg = new String(it.next(), StandardCharsets.US_ASCII).toUpperCase();
+
+            if (arg.equals("GET")) {
+               withReturnPrevious();
+               continue;
+            }
+
+            if (arg.equals("NX")) {
+               if (operationType != SetOperationType.PUT) throw new IllegalArgumentException("NX and XX are exclusive");
+               withOperationType(SetOperationType.PUT_IF_ABSENT);
+               continue;
+            }
+
+            if (arg.equals("XX")) {
+               if (operationType != SetOperationType.PUT) throw new IllegalArgumentException("NX and XX are exclusive");
+               withOperationType(SetOperationType.PUT_IF_PRESENT);
+               continue;
+            }
+
+            if (arg.charAt(0) == 'E' || arg.charAt(0) == 'P') {
```

Comment:
Same here EX, PX, EXAT, PXAT and KEEPTL are all mutually exclusive. The maximum number of arguments that can be passed is 5 across all arguments. Just doing a simple size check and then basing equality checks on that will be a ton simpler. For example if 5 is present, then we know the 3rd argument (list.get(2)) must be NX or XX and that 4th argument has to get GET etc.

Reply:
The position of the options can change, for example, all are acceptable:

```text
127.0.0.1:6379> SET key value GET NX EX 30
(nil)
127.0.0.1:6379> GET key
"value"
127.0.0.1:6379> SET key value2 XX EX 45 GET
"value"
```

Reply:
Great documentation they have 🤕

Reply:
Ah, okay, We should still be able to check them based on byte[] and length of the array. For example if the argument is 2 bytes long we know we only have to do Arrays.equals check to NX or EX.